pipelines/rag_pipelines.py

```diff
@@ -0,0 +1,109 @@
+from typing import Union, List
+
+import pandas as pd
+from langchain_core.language_models import BaseChatModel
+from langchain_core.output_parsers import StrOutputParser
+from langchain_core.prompts import ChatPromptTemplate
+from langchain_core.retrievers import BaseRetriever
+from langchain_core.vectorstores import VectorStore
+
+from retrievers.retrievers import SQLRetriever, AutoRetriever
+from langchain_core.runnables import RunnableParallel, RunnablePassthrough, RunnableSerializable
+from langchain.docstore.document import Document
+
+from settings import DEFAULT_LLM
+
+
+class LangChainRAGPipeline:
+    """
+    Builds a RAG pipeline using langchain LCEL components
+    """
+
+    def __init__(self, retriever, prompt_template, llm = DEFAULT_LLM):
+
+        self.retriever = retriever
+        self.prompt_template = prompt_template
+        self.llm = llm
+
+    def rag_with_returned_sources(self) -> RunnableSerializable:
+        """
+        Builds a RAG pipeline with returned sources
+        :return:
+        """
+
+        def format_docs(docs):
+            return "\n\n".join(doc.page_content for doc in docs)
+
+        prompt = ChatPromptTemplate.from_template(self.prompt_template)
+
+        rag_chain_from_docs = (
+                RunnablePassthrough.assign(context=(lambda x: format_docs(x["context"])))
+                | prompt
+                | self.llm
+                | StrOutputParser()
+        )
+
+        rag_chain_with_source = RunnableParallel(
+            {"context": self.retriever, "question": RunnablePassthrough()}
+        ).assign(answer=rag_chain_from_docs)
+
+        return rag_chain_with_source
+
+    @classmethod
+    def from_retriever(cls, retriever: BaseRetriever, prompt_template: str, llm: BaseChatModel):
+        """
+        Builds a RAG pipeline with returned sources using a BaseRetriever
+        :param retriever: BaseRetriever
+        :param prompt_template: str
+        :param llm: BaseChatModel
+
+        :return:
+        """
+        return cls(retriever, prompt_template, llm)
+
+    @classmethod
+    def from_sql_retriever(cls,
+                           connection_dict,
+                           retriever_prompt_template: dict,
+                           rag_prompt_template,
+                           llm: BaseChatModel = None
+                           ):
+        """
+        Builds a RAG pipeline with returned sources using a SQLRetriever
+
+        :param connection_dict:
+        :param retriever_prompt_template:
+        :param rag_prompt_template:
+        :param llm:
+
+        :return:
+        """
+        retriever = SQLRetriever(connection_dict=connection_dict, prompt_template=retriever_prompt_template)
+        return cls(retriever, rag_prompt_template, llm)
```

Comment:
Don't believe this will work as is, not had a chance to properly test yet. 

Its technically a chain and not a retreiver.